src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,496 @@
+{
+  /** length of the command */
+  unsigned int command_length = 0;
+
+  /** convert the active-frequency to an AR62xx-ID */
+  uint16_t active_frequency_idx = FrequencyToAR62xxId(active_frequency);
+
+  /** convert the passive-frequency to an AR62xx-ID */
+  uint16_t passive_frequency_idx = FrequencyToAR62xxId(passive_frequency);
+
+  /** command-start */
+  command[command_length++] = 0xA5;
+
+  /** add prot-ID */
+  command[command_length++] = 0x14;
+  command[command_length++] = 5;
+
+  /** set active station */
+  if (active_passive == ACTIVE_STATION)
+    active_frequency_idx = FrequencyToAR62xxId(frequency);
+
+  /** set passive station */
+  else
+    passive_frequency_idx = FrequencyToAR62xxId(frequency);
+
+  /** setting frequencies */
+  command[command_length++] = 22;
+  command[command_length++] = GetNthByte(active_frequency_idx, 1);
+  command[command_length++] = GetNthByte(active_frequency_idx, 0);
+  command[command_length++] = GetNthByte(passive_frequency_idx, 1);
+  command[command_length++] = GetNthByte(passive_frequency_idx, 0);
+
+  /** Creating the binary value */
+  uint16_t command_crc = CRCBitwise(command, command_length);
+  command[command_length++] = GetNthByte(command_crc, 1);
+  command[command_length++] = GetNthByte(command_crc, 0);
+  return command_length;
+}
+
+bool AR62xxDevice::ParseReceivedData(const uint8_t *data, size_t len)
+{
+  size_t idx = 0;
+  uint16_t crc;
+  uint16_t rec_buf_len = 0;
+  int command_length;
+  int rec_bufsize = 127;
+  uint8_t command[rec_bufsize];
+
+  /** loop as long we have data */
+  while (idx < len)
+  {
+    /** check if we received a header*/
+    if (data[idx] == 0xA5)
+      /** reset buffer length */
+      rec_buf_len = 0;
+
+    /** check if the buffer-size is reached */
+    if (rec_buf_len >= rec_bufsize)
+      /** reset buffer length */
+      rec_buf_len = 0;
+
+    /** write command in buffer */
+    command[rec_buf_len++] = (char)data[idx++];
+
+    /** if we received data we have to find the start of the protocol
+     *  for this we truncate the message until we find the start
+     *  of the protocol which is 0x14 (PROT-ID). For this we have
+     * a look at the first two bytes of the command.
+     * If we find this id we go on and try to parse the data below.
+    */
+    if (rec_buf_len == 2)
+    {
+      /** 
+       * check if command is NOT PROT-ID 
+       * 0x14 is the start of the PROT
+       */
+      if (!(command[rec_buf_len - 1] == 0x14))
+      {
+        /** reset buffer length */
+        rec_buf_len = 0;
+      }
+    }
+
+    /** we got some data and the data begins with the PROT-ID */
+    else if (rec_buf_len >= 3)
+    {
+      command_length = command[2];
+
+      /** all received */
+      if (rec_buf_len >= (command_length + 5))
+      {
+        /** generate uint16_t out of commands */
+        uint16_t crc_value = BytesToUnsigned(command[command_length + 4], command[command_length + 3]);
```

Comment:
What an obscure and complicated way to read a little-endian 16 bit integer!

Reply:
can you give me a hint about how to do better?

Reply:
`ReadUnalignedLE16()`.
Your code is obscure, and that hides the fact that all of this complexity and obscurity is about reading 16 bit little-endian. Using the function `ReadUnalignedLE16()` clearly expresses that in a very simple and trivial way.

Reply:
ok, thanks, i'll have a look - maybe we can throw away this piece of code and replace it with yours :-)

Reply:
replaced :-)

Reply:
and tested with live-system